src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs

```diff
@@ -0,0 +1,691 @@
+
+#nullable enable
+
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    public class FunctionPointerTests : CompilingTestBase
+    {
+        private CSharpCompilation CreateCompilationWithFunctionPointers(string source)
+        {
+            return CreateCompilation(source, options: TestOptions.UnsafeReleaseDll, parseOptions: TestOptions.RegularPreview);
+        }
+
+        private CompilationVerifier CompileAndVerifyFunctionPointers(CSharpCompilation compilation)
+        {
+            return CompileAndVerify(compilation, verify: Verification.Skipped);
+        }
+
+        [Fact]
+        public void ImplicitConversionToVoid()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    void M1(delegate*<void> p1)
+    {
+        void* v1 = p1;
+        delegate*<string> p2 = &M4;
+        M2(p2);
+        void* v2 = M3();
+    }
+
+    void M2(void* v1) {}
+    delegate*<string, int> M3() => throw null;
+    static string M4() => throw null;
+}");
+
+            var verifier = CompileAndVerifyFunctionPointers(comp);
+            verifier.VerifyIL("C.M1", expectedIL: @"
+{
+  // Code size       24 (0x18)
+  .maxstack  2
+  .locals init (void* V_0, //v1
+                delegate*<string> V_1, //p2
+                void* V_2) //v2
+  IL_0000:  ldarg.1
+  IL_0001:  stloc.0
+  IL_0002:  ldftn      ""string C.M4()""
+  IL_0008:  stloc.1
+  IL_0009:  ldarg.0
+  IL_000a:  ldloc.1
+  IL_000b:  call       ""void C.M2(void*)""
+  IL_0010:  ldarg.0
+  IL_0011:  call       ""delegate*<string,int> C.M3()""
+  IL_0016:  stloc.2
+  IL_0017:  ret
+}
+            ");
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var initializer1 = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().First().Initializer!.Value;
+            assertResult(model, initializer1);
+            var parameter = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().First().ArgumentList.Arguments.Single();
+            assertResult(model, parameter.Expression);
+            var initializer2 = tree.GetRoot().DescendantNodes().OfType<VariableDeclaratorSyntax>().Last().Initializer!.Value;
+            assertResult(model, initializer2);
+
+            static void assertResult(SemanticModel model, ExpressionSyntax initializer1)
+            {
+                var typeInfo = model.GetTypeInfo(initializer1);
+                Assert.True(typeInfo.ConvertedType is IPointerTypeSymbol { PointedAtType: { SpecialType: SpecialType.System_Void } });
+                Assert.Equal(TypeKind.FunctionPointer, typeInfo.Type!.TypeKind);
+                var conversion = model.GetConversion(initializer1);
+                Assert.Equal(ConversionKind.PointerToVoid, conversion.Kind);
```

Comment:
>conversion.Kind [](start = 59, length = 15)

Please assert IsPointer and Isimplicit too.

Reply:
I can do that, but I didn't change any of the internal state of this ConversionKind so I didn't consider it relevant to the test.

Reply:
> I can do that, but I didn't change any of the internal state of this ConversionKind so I didn't consider it relevant to the test. Let's pretend we don't know how the properties are implemented and we are testing scenario that involves conversion of a function pointer.

---
In reply to: [402658697](https://github.com/dotnet/roslyn/pull/43007#discussion_r402658697) [](ancestors = 402658697)